homeassistant/components/ring/__init__.py

```diff
@@ -146,98 +156,141 @@ def token_updater(token):
+
+
+class DeviceDataUpdater:
+    """Data storage for device data."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        config_entry_id: str,
+        ring: Ring,
+        update_method: str,
+        update_interval: timedelta,
+    ):
+        """Initialize device data updater."""
+        self.hass = hass
+        self.config_entry_id = config_entry_id
+        self.ring = ring
+        self.update_method = update_method
+        self.update_interval = update_interval
         self.devices = {}
         self._unsub_interval = None
 
-    async def track_device(self, config_entry_id, device):
+    async def track_device(self, device, update_callback):
         """Track a device."""
         if not self.devices:
             self._unsub_interval = async_track_time_interval(
-                self.hass, self.refresh_all, SCAN_INTERVAL
+                self.hass, self.refresh_all, self.update_interval
             )
 
-        key = (config_entry_id, device.device_id)
-
-        if key not in self.devices:
-            self.devices[key] = {
+        if device.device_id not in self.devices:
+            self.devices[device.device_id] = {
                 "device": device,
-                "count": 1,
+                "update_callbacks": [update_callback],
+                "data": None,
             }
+            # Store task so that other concurrent requests can wait for us to finish and
+            # data be available.
+            self.devices[device.device_id]["task"] = asyncio.current_task()
+            self.devices[device.device_id][
+                "data"
+            ] = await self.hass.async_add_executor_job(self.update_method, device)
+            self.devices[device.device_id].pop("task")
         else:
-            self.devices[key]["count"] += 1
+            self.devices[device.device_id]["update_callbacks"].append(update_callback)
+            # If someone is currently fetching data as part of the initialization, wait for them
+            if "task" in self.devices[device.device_id]:
+                await self.devices[device.device_id]["task"]
 
-        await self.hass.async_add_executor_job(device.update_health_data)
+        update_callback(self.devices[device.device_id]["data"])
 
     @callback
-    def untrack_device(self, config_entry_id, device):
+    def untrack_device(self, device, update_callback):
         """Untrack a device."""
-        key = (config_entry_id, device.device_id)
-        self.devices[key]["count"] -= 1
+        self.devices[device.device_id]["update_callbacks"].remove(update_callback)
 
-        if self.devices[key]["count"] == 0:
-            self.devices.pop(key)
+        if not self.devices[device.device_id]["update_callbacks"]:
+            self.devices.pop(device.device_id)
 
         if not self.devices:
             self._unsub_interval()
             self._unsub_interval = None
 
-    def refresh_all(self, _):
+    def refresh_all(self, _=None):
```

Comment:
Why is this not an async interface? We seem to be calling this from async context and two out of three calls inside are async.

Reply:
Jumping into the executor 10 times inside a loop is a lot of overhead of jumping around between async and sync. It's in that case better to run in sync and just schedule the callbacks to be called.